### PR TITLE
RFC: allow IOBuffer to be parameterized and remove start/stop_reading

### PR DESCRIPTION
**Pull request by vtjnash**
ref #9888
ref #1925

~~(note: these two pull requests have gotten mixed up together in my local repository commits. i'll definitely fix that and sort them into proper commits before merging this)~~

### CODE REVIEW

File: base/build.h [hunk 1]
```diff
@@ -1,23 +0,0 @@
-// This file is a part of Julia. License is MIT: http://julialang.org/license
```
Comment (tkelman) [inline base/build.h]: this deletion looks unrelated, missed from a0e503ecec904e4ebc34e6529baa5912ed64ea61 ?

Comment (vtjnash) [inline base/build.h]: yeah, looks like my branches got mixed up in the rebase

**Comment by JeffBezanson** [comment]
:+1: :100: to the start_reading change.

The IOBuffer change seems like an awful lot of effort just to support read-only IOBuffers of substrings of ASCII or UTF-8. I haven't seen sufficient motivation for that.

File: base/stream.jl [hunk 2]
```diff
@@ -307,26 +331,24 @@ end
...
 function wait_readnb(x::AsyncStream, nb::Int)
     while isopen(x) && nb_available(x.buffer) < nb
-        start_reading(x)
-        stream_wait(x,x.readnotify)
+        stream_waitn(x, nb - nb_available(x.buffer), x.readnotify)
     end
 end
 
-wait_close(x) = if isopen(x) stream_wait(x,x.closenotify); end
+wait_close(x) = wait_readnb(x, typemax(Int))
```
Comment (vtjnash) [inline base/stream.jl]: self note: revert this line

File: base/stream.jl [hunk 3]
```diff
@@ -307,26 +331,24 @@ end
...
-        stream_wait(x,x.readnotify)
+        stream_waitn(x, 1, x.readnotify)
     end
 end
 
 function wait_readnb(x::AsyncStream, nb::Int)
     while isopen(x) && nb_available(x.buffer) < nb
-        start_reading(x)
-        stream_wait(x,x.readnotify)
+        stream_waitn(x, nb - nb_available(x.buffer), x.readnotify)
```
Comment (vtjnash) [inline base/stream.jl]: self note: should be just `nb`

**Comment by vtjnash** [comment]
> The IOBuffer change seems like an awful lot of effort just to support read-only IOBuffers of substrings of ASCII or UTF-8. I haven't seen sufficient motivation for that.

I really need to go through and split up these commits into bug fixes and changes. The actual parameterization of the IOBuffer is done in very few changes (most just changing of typed fields from IOBuffer to SimpleIOBuffer). The rest is either unrelated (but got mixed up in my local branches), or bugfixes / code simplifications.

I think the more useful motivating case for this is being able to make IOBuffers backed by SubArrays or SharedArrays.

**Comment by JeffBezanson** [comment]
I wouldn't underestimate the annoyance of having both `IOBuffer` and `SimpleIOBuffer`. Although so far I'm still against the change, maybe the renaming should be reversed, to `AnyIOBuffer` and `IOBuffer`.

**Comment by ScottPJones** [comment]
I don't know, I think this *might* be quite useful for certain types of operations, say where you've got data streaming into buffers, a buffer is filled, and then you start filling another buffer... normally trying to process the data in those buffers is a pain, but this would hide that, I think... and could be a big win for the sorts of stuff that I do.  I don't care what the name is, just that the functionality is there, and that use can use these anywhere you'd read from an IOBuffer currently (without causing type instability or performance issues for code that *doesn't* use the new type IOBuffers...)

**Comment by vtjnash** [comment]
@JeffBezanson i've cleaned up the commits and done the renaming you suggested. i agree that is better. I initially though the abstract type would be useful for clients, but now realize that was largely false, since the primarily useful abstract type for annotation would be `IO`.

**Comment by vtjnash** [comment]
travis is now ok with this (modulo the OOM death on one of the attempts). any final thoughts before merging?

File: base/build.h (hunk 1 above)
Comment (kmsquire) [inline base/build.h]: Still needs a rebase here?

Comment (tkelman) [inline base/build.h]: I think strictly speaking no, just this file deletion could have been committed to master by itself at any time.

**Comment by ScottPJones** [comment]
FWIW, LGTM :+1:

**Comment by ScottPJones** [comment]
Argh: spoke too soon, but this still can be merged and I can submit a PR later for the minor thing I just noticed.
It outputs a replacement character `'\ufffd'` if the `Char` is `> 0x10ffff`, but it doesn't check the surrogate range, which are also invalid, and should be replaced by `'\ufffd'`.
